### PR TITLE
API Fixes

### LINKED ISSUE #3885: 10.7.0 nightly: TV programms: Exception: Guid can't be empty (Parameter 'id')
**Describe the bug**

**System (please complete the following information):**
 - OS: [e.g. Debian, Windows] Linux Ubuntu
 - Virtualization: [e.g. Docker, KVM, LXC] Docker
 - Clients: [Browser, Android, Fire Stick, etc.] Browser
 - Browser: [e.g. Firefox 72, Chrome 80, Safari 13] Chromium, Firefox
 - Jellyfin Version: [e.g. 10.4.3, nightly 20191231] Todays 10.7.0 nightly
 - Playback: [Direct Play, Remux, Direct Stream, Transcode] 
 - Installed Plugins: [e.g. none, Fanart, Anime, etc.]
 - Reverse Proxy: [e.g. none, nginx, apache, etc.] nginx
 - Base URL: [e.g. none, yes: /example]
 - Networking: [e.g. Host, Bridge/NAT]
 - Storage: [e.g. local, NFS, cloud]

**To Reproduce**

1. Update to todays 10.7.0 nightly
2. Open Live tv programms list
3. See error in log and list doesn't open

**Expected behavior**

Live tv programms table should be shown

**Logs**

```
[22:31:31] [ERR] [33] Jellyfin.Server.Middleware.ExceptionMiddleware: Error processing request. URL GET /LiveTv/Programs.                                                                │
│   System.ArgumentException: Guid can't be empty (Parameter 'id')                                                                                                                           │
│      at Emby.Server.Implementations.Library.LibraryManager.GetItemById(Guid id)                                                                                                            │
│      at Jellyfin.Api.Controllers.LiveTvController.GetLiveTvPrograms(String channelIds, Nullable`1 userId, Nullable`1 minStartDate, Nullable`1 hasAired, Nullable`1 isAiring, Nullable`1    │
│   maxStartDate, Nullable`1 minEndDate, Nullable`1 maxEndDate, Nullable`1 isMovie, Nullable`1 isSeries, Nullable`1 isNews, Nullable`1 isKids, Nullable`1 isSports, Nullable`1 startIndex,   │
│    Nullable`1 limit, String sortBy, String sortOrder, String genres, String genreIds, Nullable`1 enableImages, Nullable`1 imageTypeLimit, String enableImageTypes, Nullable`1 enableUser   │
│   Data, String seriesTimerId, Nullable`1 librarySeriesId, String fields, Boolean enableTotalRecordCount)                                                                                   │
│      at lambda_method(Closure , Object )                                                                                                                                                   │
│      at Microsoft.Extensions.Internal.ObjectMethodExecutorAwaitable.Awaiter.GetResult()                                                                                                    │
│      at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object control   │
│   ler, Object[] arguments)                                                                                                                                                                 │
│      at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)       │
│      at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope s   │
│   cope, Object state, Boolean isCompleted)                                                                                                                                                 │
│      at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)                                                                       │
│      at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)                                               │
│      at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()                                                                                           │
│   --- End of stack trace from previous location where exception was thrown ---                                                                                                             │
│      at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state,   │
│    Boolean isCompleted)                                                                                                                                                                    │
│      at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)                                                                             │
│      at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)                                                       │
│      at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()                                                                                                │
│   --- End of stack trace from previous location where exception was thrown ---                                                                                                             │
│      at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)                                        │
│      at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)                                               │
│      at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)                                                                                             │
│      at Swashbuckle.AspNetCore.ReDoc.ReDocMiddleware.Invoke(HttpContext httpContext)                                                                                                       │
│      at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)                                                                                               │
│      at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)                                                                 │
│      at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)                                                                                           │
│      at Microsoft.AspNetCore.ResponseCompression.ResponseCompressionMiddleware.Invoke(HttpContext context)                                                                                 │
│      at Jellyfin.Server.Middleware.ResponseTimeMiddleware.Invoke(HttpContext context)
```

**Screenshots**

**Additional context**

TV programms table works on 10.6.0 stable but no longer on 10.7.0 nightly (regression).

### PR DESCRIPTION
**Pull request by crobibero**
95% copy & paste from controller to helper class.

Required attribute here was causing validation error responses to web, guess the parameter wasn't actually required.
https://github.com/jellyfin/jellyfin/pull/3861/commits/7b864e3a3c286c9502b6ef3ff1d13754e796706a

Also snuck in some api route fixes.

Fixes #3885

### CODE REVIEW

File: Jellyfin.Api/Helpers/DynamicHlsHelper.cs [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+            BaseJellyfinApiController controller,
+            StreamingRequestDto streamingRequest,
+            bool isHeadRequest,
+            bool enableAdaptiveBitrateStreaming,
+            TranscodingJobType transcodingJobType,
+            CancellationTokenSource cancellationTokenSource)
+        {
+            using var state = await StreamingHelpers.GetStreamingState(
+                    streamingRequest,
+                    controller.Request,
```
Comment (cvium) [inline Jellyfin.Api/Helpers/DynamicHlsHelper.cs:133]: It seems this code is only ever accessing Request and Response. Both of which are available via `IHttpContextAccessor`. Please add it as a constructor argument and remove the dependency on `BaseJellyfinApiController`.

Since all three helpers are Scoped, `IHttpContextAccessor` will always contain the _current_ HttpContext.

File: Jellyfin.Api/Controllers/PlaystateController.cs [hunk 2]
```diff
@@ -73,15 +74,21 @@ public class PlaystateController : BaseJellyfinApiController
         public ActionResult<UserItemDataDto> MarkPlayedItem(
             [FromRoute] Guid userId,
             [FromRoute] Guid itemId,
-            [FromQuery] DateTime? datePlayed)
+            [FromQuery] string? datePlayed)
         {
+            DateTime? parsedDatePlayed = null;
+            if (!string.IsNullOrEmpty(datePlayed))
+            {
+                parsedDatePlayed = DateTime.ParseExact(datePlayed, "yyyyMMddHHmmss", CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal);
```
Comment (cvium) [inline Jellyfin.Api/Controllers/PlaystateController.cs:82]: Would it make sense to do a Type Converter instead? https://stackoverflow.com/a/52822295

Comment (crobibero) [reply Jellyfin.Api/Controllers/PlaystateController.cs:82]: learn something new every day!

File: Jellyfin.Api/Controllers/PlaystateController.cs [hunk 3]
```diff
@@ -1,4 +1,5 @@
 ﻿using System;
+using System.ComponentModel;
```
Comment (cvium) [inline Jellyfin.Api/Controllers/PlaystateController.cs:2]: Why is this needed now?

File: Jellyfin.Api/Controllers/UniversalAudioController.cs [hunk 4]
```diff
@@ -95,24 +110,68 @@ public class UniversalAudioController : BaseJellyfinApiController
...
-                mediaSourceId,
-                null,
-                new DeviceProfileDto { DeviceProfile = deviceProfile })
+            var authInfo = _authorizationContext.GetAuthorizationInfo(Request);
+
+            _logger.LogInformation("GetPostedPlaybackInfo profile: {@Profile}", deviceProfile);
+
+            if (deviceProfile == null)
+            {
+                var caps = _deviceManager.GetCapabilities(authInfo.DeviceId);
```
Comment (cvium) [inline Jellyfin.Api/Controllers/UniversalAudioController.cs:122]: was this copied from somewhere? Not liking the variable name.

File: Jellyfin.Api/Helpers/MediaInfoHelper.cs [hunk 5]
```diff
@@ -0,0 +1,573 @@
...
+
+            var result = await _mediaSourceManager.OpenLiveStream(request, CancellationToken.None).ConfigureAwait(false);
+
+            var profile = request.DeviceProfile;
+            if (profile == null)
+            {
+                var caps = _deviceManager.GetCapabilities(authInfo.DeviceId);
+                if (caps != null)
+                {
+                    profile = caps.DeviceProfile;
```
Comment (cvium) [inline Jellyfin.Api/Helpers/MediaInfoHelper.cs:471]:
```suggestion
                var clientCapabilities = _deviceManager.GetCapabilities(authInfo.DeviceId);
                if (clientCapabilities != null)
                {
                    profile = clientCapabilities.DeviceProfile;
```

**Review by cvium — APPROVED** [review]
